### PR TITLE
V3.0.0 beta development

### PR DESCRIPTION
**Pull request by sanjeevbigstep**
- Admin Chat Settings with chat enable/disable option for Logged-in users.
- Admin Colors Settings.
- Added UI for auto-play audio error. (Working on its UI).
- Active Speaker.
- Ghost Mode.
- User avatar is shown on Video Mute (of Logged-in users).

### CODE REVIEW

**Comment by geotraveler** [comment]
Just came here to add that watching all these merges & commits is exciting as hell! Can't wait for Version 3.0 👍

**Review by digitallysavvy — CHANGES_REQUESTED** [review]
What is the purposes of holding all streams in an Array, this is the least flexible data type. the local and remote streams are already there. Please use the existing windows variables and do not use this array, it is not the correct data type. if you wish to move all streams into a single variable, use an object that encompasses the two existing objects of remote and local streams

**Review by digitallysavvy — CHANGES_REQUESTED** [review]
Agora provides an event listener for active speaker there is no reason to create this logic from scratch.

**Comment by sanjeevbigstep** [comment]
> What is the purposes of holding all streams in an Array, this is the least flexible data type. the local and remote streams are already there. Please use the existing windows variables and do not use this array, it is not the correct data type. if you wish to move all streams into a single variable, use an object that encompasses the two existing objects of remote and local streams

Hi Hermes ,
The main purpose of combining these streams into one variable was -
- To store the user details like profile pic/user name only one time and store them in this variable. (So that we don't need to call ajax every time when user is going to mute their video - It will store the user's data when we are adding remote streams).
- The second main purpose was that it can be used at other places as well like in chatting, we can directly use this variable to get the details instead of calling ajax (to get user's profile pic).
- The last one was to merging all streams into one.

Please let me know if needs to remove this or any other improvements.

**Review by digitallysavvy — APPROVED** [review]

**Review by digitallysavvy — COMMENTED** [review]
this looks good

File: includes/class-wp-agora-io-channel.php [hunk 1]
```diff
@@ -47,7 +47,8 @@ class WP_Agora_Channel {
     'region' => 0,
     'bucket' => '',
     'accessKey' => '',
-    'secretKey' => ''
+    'secretKey' => '',
+    'chat-support-loggdin' => '',
```
Comment (digitallysavvy) [inline includes/class-wp-agora-io-channel.php]: what's this key for?

File: admin/views/agora-admin-new-channel.php [hunk 2]
```diff
@@ -434,24 +434,24 @@ function render_agoraio_channel_form_chat_support($channel) {
...
-        if(channel_type == 'communication'){
-          jQuery('#pre_call_video').parents('tr').css('display','block');
-        }else{
-          jQuery('#pre_call_video').parents('tr').css('display','none');
-        }
-      }
-      jQuery('#type').change(ShowHideFields);
-      jQuery('#type').change();
-    });
+    // window.addEventListener('load', function() {
```
Comment (digitallysavvy) [inline admin/views/agora-admin-new-channel.php:437]: why has this section been commented out?

File: public/class-wp-agora-io-public.php [hunk 3]
```diff
@@ -375,6 +375,8 @@ public function enqueue_scripts() {
 
 		wp_enqueue_script( $this->plugin_name, plugin_dir_url( __FILE__ ) . 'js/wp-agora-io-public.js', array( 'jquery' ), $this->version, false );
 
+		wp_enqueue_script( $this->plugin_name.'-agora-deviceTest-js', plugin_dir_url( __FILE__ ) . 'js/wp-agora-io-device-test.js', array( 'jquery' ), $this->version, false );
```
Comment (digitallysavvy) [inline public/class-wp-agora-io-public.php:378]: are we loading this script every time? this should only get enqueued if the pre-call is enabled.

File: public/js/agora-broadcast-client.js [hunk 4]
```diff
@@ -131,21 +131,25 @@ async function createCameraStream(uid, deviceIds) {
...
     /* Mute Audios and Videos Based on Mute All Users Settings */
-    if(window.mute_all_users_audio_video){
-      if(localStream.getVideoTrack() && localStream.getVideoTrack().enabled){
-        jQuery("#video-btn").trigger('click');
-      }
-      if(localStream.getAudioTrack() && localStream.getAudioTrack().enabled){
-        jQuery("#mic-btn").trigger('click');
-      }
-    }
+    // if(window.mute_all_users_audio_video){
```
Comment (digitallysavvy) [inline public/js/agora-broadcast-client.js:145]: why has this been commented out?

File: public/js/wp-agora-io-public.js [hunk 5]
```diff
@@ -406,7 +440,8 @@ window.AGORA_UTILS = {
   },
 
   // REMOTE STREAMS UI
-  addRemoteStreamView: function(remoteStream) {
+  addRemoteStreamView: function(remoteStream, cond='') {
+    if(!window.isPublished && window.agoraMode!="audience") return;
```
Comment (digitallysavvy) [inline public/js/wp-agora-io-public.js:444]: what is the purpose of this `if` statement?

File: public/js/wp-agora-io-public.js [hunk 6]
```diff
@@ -449,19 +497,41 @@ window.AGORA_UTILS = {
...
       if ((err && err.status !== "aborted") || (err && err.audio && err.audio.status !== "aborted")){
+        console.log("hnjiErrorDuringPlay")
         jQuery('body #' + streamId + '_container').prepend(
           addAudioErrorGesture(streamId)
         )
       }
 
       handleGhostMode(streamId, 'remote');
 
+      if(cond=='playAfterPublish'){
```
Comment (digitallysavvy) [inline public/js/wp-agora-io-public.js:514]: please explain this logic

File: public/js/wp-agora-io-public.js [hunk 7]
```diff
@@ -748,7 +818,9 @@ jQuery(document).ready(function(){
   const THRESHOLD_AUDIO_LEVEL = 0.1;
   setInterval(() => {
 
-    window.allStreams = Object.fromEntries(Object.entries(window.allStreams).filter(([_, v]) => v != null));
+    if(typeof window.allStreams!='undefined'){
+      window.allStreams = Object.fromEntries(Object.entries(window.allStreams).filter(([_, v]) => v != null));
```
Comment (digitallysavvy) [inline public/js/wp-agora-io-public.js:822]: what is the point of this?

**Review by digitallysavvy — CHANGES_REQUESTED** [review]

**Comment by digitallysavvy** [comment]
> > What is the purposes of holding all streams in an Array, this is the least flexible data type. the local and remote streams are already there. Please use the existing windows variables and do not use this array, it is not the correct data type. if you wish to move all streams into a single variable, use an object that encompasses the two existing objects of remote and local streams
> 
> Hi Hermes ,
> The main purpose of combining these streams into one variable was -
> 
> * To store the user details like profile pic/user name only one time and store them in this variable. (So that we don't need to call ajax every time when user is going to mute their video - It will store the user's data when we are adding remote streams).
> * The second main purpose was that it can be used at other places as well like in chatting, we can directly use this variable to get the details instead of calling ajax.
> * The last one was to merging all streams into one.
> 
> Please let me know if we need to remove this or any other improvements.


Don't use an array that is the least efficient data type, you should be expanding upon the existing objects for local and remote streams if you need to keep track of more data on the client side. This needs to be fixed.

**Comment by digitallysavvy** [comment]
> * Admin Chat Settings with chat enable/disable option for Logged-in users.
> * Admin Colors Settings.
> * Added UI for auto-play audio error. (Working on its UI).
> * Active Speaker.
> * Ghost Mode.
> * User avatar is shown on Video Mute (of Logged-in users).

Please include any issue numbers that are being addressed within this PR to create greater visibility

File: public/js/wp-agora-io-public.js (hunk 5 above)
Comment (sanjeevbigstep) [reply public/js/wp-agora-io-public.js:444]: > what is the purpose of this `if` statement?

Hi Hermes,

As in pre-call device test, remote stream will be added by stream-added event (even when local stream is not published), so to play the remote stream only after local stream is published(After the pre-call device test is finished), if(!window.isPublished), condition is added. And to play the same after pre call device is completed, called the same function addRemoteStream when stream is published.

File: public/class-wp-agora-io-public.php (hunk 3 above)
Comment (sanjeevbigstep) [reply public/class-wp-agora-io-public.php:378]: > are we loading this script every time? this should only get enqueued if the pre-call is enabled.

Thanks. Will do that in next PR.

File: public/js/agora-broadcast-client.js (hunk 4 above)
Comment (sanjeevbigstep) [reply public/js/agora-broadcast-client.js:145]: > why has this been commented out?

This should be work after stream is published (not in pre-call device test, so moved this code in publish stream function.), was testing this condition, will remove it in next PR (from commented section).

File: public/js/agora-communication-client.js [hunk 8]
```diff
@@ -95,8 +101,7 @@ async function isVideoAvailable() {
   return devices.some(device => 'videoinput' === device.kind);
 }
 
-// video streams for channel
-function createCameraStream(uid, next) {
+function createTmpCameraStream(uid, next){
```
Comment (digitallysavvy) [inline public/js/agora-communication-client.js:104]: Why is this method created twice? the `communication-client.js` and `broadcast-client.js` should only contain unique methods. Reading through the code this looks identical to the `broadcast-client.js`. Common  methods should go into `wp-agora-io-public.js`

File: admin/class-wp-agora-io-admin.php [hunk 9]
```diff
@@ -42,8 +42,12 @@ public function saveAjaxSettings() {
 		unset($_POST['action']);
 		$keys = array_keys($_POST);
 		$key = $keys[0];
-		$value = sanitize_text_field( $_POST[$key] );
-
+		if($key != 'global_colors'){
+			$value = sanitize_text_field( $_POST[$key] );
+		}else{
+			$value = $_POST[$key];
```
Comment (digitallysavvy) [inline admin/class-wp-agora-io-admin.php:48]: why isn't this input sanitized?

File: includes/class-wp-agora-io-channel.php [hunk 10]
```diff
@@ -239,11 +284,17 @@ public function save( $args ) {
       $recordingSettings[$key] = sanitize_text_field($args[$key]);
       return $recordingSettings[$key];
     }, array_keys(self::$defaultRecordingSettings));
-
+    
     update_post_meta($post_id, 'channel_video_settings', $videoSettings);
```
Comment (digitallysavvy) [inline includes/class-wp-agora-io-channel.php:288]: Why aren't these inputs sanitized? all the new inputs are sanitized, why not update the old ones?

File: public/class-wp-agora-io-public.php [hunk 11]
```diff
@@ -57,6 +76,127 @@ public function __construct( $plugin_name, $version ) {
...
+
+		wp_die();
+	}
+
+	/* Function to get chats from the databse */
+	public function getChatsFromHistory(){
+		global $wpdb;
+		$table_name = $wpdb->prefix . 'agora_io_chats';
+
+		$channel_id = $_POST['channel_id'];
```
Comment (digitallysavvy) [inline public/class-wp-agora-io-public.php:93]: values should be sanitized before they are uses for DB queries.

File: public/class-wp-agora-io-public.php [hunk 12]
```diff
@@ -57,6 +76,127 @@ public function __construct( $plugin_name, $version ) {
...
+		}
+		echo json_encode($results);
+		wp_die();
+	}
+
+	/* Function to save chat in the database if chat is enabled */
+	public function saveChat(){
+		global $wpdb;
+
+		$channel_id = $_POST['channel_id'];
```
Comment (digitallysavvy) [inline public/class-wp-agora-io-public.php:123]: Input values must be sanitized

**Review by digitallysavvy — CHANGES_REQUESTED** [review]
Please make sure to sanitize all inputs

**Comment by sanjeevbigstep** [comment]
> Agora provides an event listener for active speaker there is no reason to create this logic from scratch.

Fixed.

> Agora provides an event listener for active speaker there is no reason to create this logic from scratch.

Done. Kindly check.

> Agora provides an event listener for active speaker there is no reason to create this logic from scratch.



> Agora provides an event listener for active speaker there is no reason to create this logic from scratch.



> Agora provides an event listener for active speaker there is no reason to create this logic from scratch.



> Agora provides an event listener for active speaker there is no reason to create this logic from scratch.

Done. Kindly check it.

**Comment by sanjeevbigstep** [comment]
> Agora provides an event listener for active speaker there is no reason to create this logic from scratch.

Done. Kindly check it.

File: public/class-wp-agora-io-public.php (hunk 3 above)
Comment (sanjeevbigstep) [reply public/class-wp-agora-io-public.php]: Done. Kindly check it.

File: public/js/wp-agora-io-public.js (hunk 5 above)
Comment (sanjeevbigstep) [reply public/js/wp-agora-io-public.js]: > what is the purpose of this `if` statement?

Fixed. Kindly check it.

File: admin/views/agora-admin-new-channel.php (hunk 2 above)
Comment (sanjeevbigstep) [reply admin/views/agora-admin-new-channel.php]: > why has this section been commented out?

Fixed. Kindly check it.

File: public/js/agora-broadcast-client.js (hunk 4 above)
Comment (sanjeevbigstep) [reply public/js/agora-broadcast-client.js]: > > why has this been commented out?
> 
> This should be work after stream is published (not in pre-call device test, so moved this code in publish stream function.), was testing this condition, will remove it in next PR (from commented section).

Fixed. Kindly check it.

File: public/js/wp-agora-io-public.js (hunk 7 above)
Comment (sanjeevbigstep) [reply public/js/wp-agora-io-public.js]: > what is the point of this?

This was of that array that we have already removed.(to maintain local and remote streams). Kindly check it.

**Comment by sanjeevbigstep** [comment]
> > > What is the purposes of holding all streams in an Array, this is the least flexible data type. the local and remote streams are already there. Please use the existing windows variables and do not use this array, it is not the correct data type. if you wish to move all streams into a single variable, use an object that encompasses the two existing objects of remote and local streams
> > 
> > 
> > Hi Hermes ,
> > The main purpose of combining these streams into one variable was -
> > 
> > * To store the user details like profile pic/user name only one time and store them in this variable. (So that we don't need to call ajax every time when user is going to mute their video - It will store the user's data when we are adding remote streams).
> > * The second main purpose was that it can be used at other places as well like in chatting, we can directly use this variable to get the details instead of calling ajax.
> > * The last one was to merging all streams into one.
> > 
> > Please let me know if we need to remove this or any other improvements.
> 
> Don't use an array that is the least efficient data type, you should be expanding upon the existing objects for local and remote streams if you need to keep track of more data on the client side. This needs to be fixed.

Fixed. Kindly check it.

File: public/class-wp-agora-io-public.php (hunk 12 above)
Comment (sanjeevbigstep) [reply public/class-wp-agora-io-public.php]: > Input values must be sanitized

Fixed. Kindly check it.

File: public/js/agora-communication-client.js (hunk 8 above)
Comment (sanjeevbigstep) [reply public/js/agora-communication-client.js]: > Why is this method created twice? the `communication-client.js` and `broadcast-client.js` should only contain unique methods. Reading through the code this looks identical to the `broadcast-client.js`. Common methods should go into `wp-agora-io-public.js`

Fixed. Kindly check it.

**Review by digitallysavvy — APPROVED** [review]

**Comment by sanjeevbigstep** [comment]
Hi Hermes,

I found that we are muting video during screen share stop and playing the main local stream again ? What is the logic behind this ? (As it's creating error in the functionality as when we are muting video, mute-video event is getting triggered on remote side.)

**Pull request merged**